### PR TITLE
add new regression test class: parallel

### PR DESCRIPTION
**Pull request by akva2**
this compares the output of a serial and a parallel run for a simulation.

note this is just a draft. the tests all currently fail.the number of processes is hardcoded to 4 and my choice of models was just because those were the examples used earlier.

do you agree this is a good idea and if so, which models, which processes? 

in any case these models should pass i imagine. i have set very slack tolerances for the tests to point out that eventually we have differing values in integer fields. so i assume my understanding of the io is broken; i thought it was all based on serialization so ultimately we write the same data in the same order (up to rounding) from one process in either runs.

### CODE REVIEW

**Comment by atgeirr** [comment]
I think we should not expect the same result: the linear solutions are different, (also if you are comparing flow to flow_mpi the underlying grid implementation is different) bringing with it the whole "eventually they will diverge" thing. Therefore a direct comparison may be less useful. However, the solutions should be equivalent, so it is a good idea to run acceptance tests against a static known-good reference solution (which is the same for any version of flow and any number of processes) and regression tests against a regularly updated solution (which is specific to the number of processes and the program used).

**Comment by blattms** [comment]
The current master is definitely not doing that because of #892. Hopefully, this is fixed by #900.

**Comment by blattms** [comment]
You might need some kind of timeout for the parallel test as the master seems rather unstable for parallel runs. Otherwise you will eventually get some deadlocked tests hanging around.

**Comment by alfbr** [comment]
Following the current design principle for the tests, we should compare MPI runs with earlier MPI reference runs to allow strict tolerances. However, in this case we are looking at SPE1, SPE3 and SPE9, I suspect we can get by with strict tolerances even comparing MPI runs with a reference serial runs.

Generally speaking, I really welcome this PR. It is dearly needed, whether in this form or another.

> You might need some kind of timeout for the parallel test as the master seems rather unstable for parallel runs.

This is something that bites us regularly on the benchmarking. Any hope we can fix the underlying issue (i.e., ensure that MPI runs don't stall because of an exception in one of the processe)?

**Comment by akva2** [comment]
what you stated with many words i used 'up to rounding' to avoid describing. you are solving the same problem with different tools. the results obviously won't compare 1-to-1 but they should compare well within the expected single precision of the output. @blattms ctest has a default timeout. cannot recall the number but they won't be around for long (it's 10 min or something).

with #900 spe3 passes. spe1 and spe9 still fails. the problem seems to be in the well states. to give you an idea how close things are when the code actually works:
```
Comparing PRESSURE...done.
Average relative deviation = 1.88782e-05
Comparing RS...done.
Average relative deviation = 2.56151e-05
```
these numbers seems sane to me (remember data is in single precision).

**Comment by akva2** [comment]
@alfbr you are sort of adressing a different problem. my goal here is not testing for regressions in the parallel output, i'm testing for differences compared to the serial output to cover the (in)ability to do changes to io and not think about the parallel aspects of the change. i agree that the test class you describe makes sense in addition, but it does not replace the proposed test class.

**Comment by alfbr** [comment]
> cover the (in)ability to do changes to io and not think about the parallel aspects of the change.

Good point. Combining that with the kind of tolerances we can use, this PR should be able to bring down both birds with one rock.

**Comment by blattms** [comment]
> ctest has a default timeout.

Please note that this is might not kill all slave processes. I experienced stale processes on cluster nodes even with timeouts set by the queuing system.

**Comment by akva2** [comment]
spe1 and spe9 both fail for the same reason: well controls are a bunch of zeros in the parallel files. my educated guess: well controls for wells on other processes are not collected on the io processes.

**Comment by akva2** [comment]
yup. it's definitely the well state. the well state passed to output only contains local wells. it's not trivial to gather the data in correct order here. are the global wells hiding somewhere?

**Comment by akva2** [comment]
ignoring the restart data, and considering only summary the results still look too far off for my comfort. to me it looks like the parallel formulation is not solving the same problem.

**Comment by blattms** [comment]
It is known that the well information is not complete, #716 might be
related.

**Comment by akva2** [comment]
yeah. problem is i cannot find the necessary mappings to reconstruct things. i did a local hack (with some hacks on top, sprinkling in some hacks for good faith) and i can more or less reproduce acceptable OPM_[X|I]WELL data but i had to make assumptions cause there is no global well ordering available (so i hardcoded the mapping for this case). but as i say, ignoring restart files we're still only within 1e-2 rtol on the primaries in the summary file. 

digging deeper i have now adjusted some of the more interesting defaults for tolerances, and summary tests pass fine for all 3 models (1e-5 on rtol, 0.02 for atol which is a value tuned according to pressures + round-off). 

tldr; my trust in results has been restored. restart data is broken, summary is all good.

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
the two failures are to be expected as the opm-data update job hasn't popped from the queue yet (OPM/opm-data#141). the important bit it is that the parallel tests passed \o/

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
tests pass -> shameless self-merge.
